specs/_features/eip7251/beacon-chain.md

```diff
@@ -0,0 +1,907 @@
+    pending_consolidations: List[PendingConsolidation, PENDING_CONSOLIDATIONS_LIMIT]  # [New in EIP-7251]
+```
+
+#### `BeaconBlockBody`
+
+```python
+class BeaconBlockBody(Container):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data  # Eth1 data vote
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    # Execution
+    execution_payload: ExecutionPayload 
+    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
+    blob_kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
+    consolidations: List[SignedConsolidation, MAX_CONSOLIDATIONS]  # [New in EIP-7251]
+```
+
+## Helpers
+
+### Predicates
+
+#### Updated `is_eligible_for_activation_queue`
+
+```python
+def is_eligible_for_activation_queue(validator: Validator) -> bool:
+    """
+    Check if ``validator`` is eligible to be placed into the activation queue.
+    """
+    return (
+        validator.activation_eligibility_epoch == FAR_FUTURE_EPOCH
+        and validator.effective_balance >= MIN_ACTIVATION_BALANCE  # [Modified in EIP7251]
+    )
+```
+
+#### New `has_compounding_withdrawal_credential`
+
+```python
+def has_compounding_withdrawal_credential(validator: Validator) -> bool:
+    """
+    Check if ``validator`` has an 0x02 prefixed "compounding" withdrawal credential.
+    """
+    return validator.withdrawal_credentials[:1] == COMPOUNDING_WITHDRAWAL_PREFIX
+```
+
+#### New `has_execution_withdrawal_credential`
+
+```python
+def has_execution_withdrawal_credential(validator: Validator) -> bool:
+    """
+    Check if ``validator`` has a 0x01 or 0x02 prefixed withdrawal credential.
+    """
+    return has_compounding_withdrawal_credential(validator) or has_eth1_withdrawal_credential(validator)
+```
+
+#### Updated `is_fully_withdrawable_validator`
+
+```python
+def is_fully_withdrawable_validator(validator: Validator, balance: Gwei, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is fully withdrawable.
+    """
+    return (
+        has_execution_withdrawal_credential(validator)  # [Modified in EIP7251]
+        and validator.withdrawable_epoch <= epoch
+        and balance > 0
+    )
+```
+
+#### Updated `is_partially_withdrawable_validator`
+
+```python
+def is_partially_withdrawable_validator(validator: Validator, balance: Gwei) -> bool:
+    """
+    Check if ``validator`` is partially withdrawable.
+    """
+    return (
+        has_execution_withdrawal_credential(validator)  # [Modified in EIP7251]
+        and get_validator_excess_balance(validator, balance) > 0
+    )
```

Comment:
In capella spec, there is a check in the `is_partially_withdrawable_validator` method:

```python
has_max_effective_balance = validator.effective_balance == MAX_EFFECTIVE_BALANCE
```

It is needed so that a validator with effective balance < `MAX_EFFECTIVE_BALANCE` can reach the balance = `MAX_EFFECTIVE_BALANCE + 0.25 eth` without being partially withdrawn. E.g. a validator with effective balance = 31 eth must have balance more than 32.25 eth to trigger updating his effective balance to 32 eth.


```suggestion
    if has_eth1_withdrawal_credential(validator) and validator.effective_balance == MIN_ACTIVATION_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    if has_compounding_withdrawal_credential(validator) and validator.effective_balance == MAX_EFFECTIVE_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    return False
```

Reply:
Good catch! I think we should refactor the spec a bit, and make the code of the sweep similar to Cancun, something like in this commit: https://github.com/mkalinin/eth2.0-specs/commit/4c7a65ac409ba653e531142f99558972483b84d2